// @react-three/xr/dist/index.js
"use strict";
Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: "Module" } });
const Controllers = require("./Controllers.js");
const Hands = require("./Hands.js");
const Interactions = require("./Interactions.js");
const XR = require("./XR.js");
const XRController = require("./XRController.js");
const XREvents = require("./XREvents.js");
exports.Controllers = Controllers.Controllers;
exports.Ray = Controllers.Ray;
exports.Hands = Hands.Hands;
exports.InteractionManager = Interactions.InteractionManager;
exports.Interactive = Interactions.Interactive;
exports.RayGrab = Interactions.RayGrab;
exports.useHitTest = Interactions.useHitTest;
exports.useInteraction = Interactions.useInteraction;
exports.ARButton = XR.ARButton;
exports.VRButton = XR.VRButton;
exports.XR = XR.XR;
exports.XRButton = XR.XRButton;
exports.useController = XR.useController;
exports.useXR = XR.useXR;
exports.XRController = XRController.XRController;
exports.useXREvent = XREvents.useXREvent;
